poc_nextcloud.services.yml

```diff
@@ -0,0 +1,35 @@
+services:
+  _defaults:
+    autowire: true
+  logger.channel.poc_nextcloud:
+    parent: logger.channel_base
+    arguments: ['poc_nextcloud']
+  # Services that depend on the nextcloud connection will cause exception if
+  # Nextcloud is not configured.
+  Drupal\poc_nextcloud\Connection\ApiConnectionInterface:
+    factory: [Drupal\poc_nextcloud\Connection\ApiConnection, 'fromConfig']
+    arguments:
+      - '@http_client'
+      - '@config.factory'
+      - '@state'
+  Drupal\poc_nextcloud\Service\NextcloudUserMap: {}
+  Drupal\poc_nextcloud\Service\NextcloudUrlBuilder:
+    factory: [~, create]
+    arguments: ['@config.factory']
+  Drupal\poc_nextcloud\Endpoint\NxUserEndpoint: {}
+  Drupal\poc_nextcloud\Endpoint\NxGroupEndpoint: {}
+  Drupal\poc_nextcloud\Endpoint\NxGroupFolderEndpoint: {}
+  Drupal\poc_nextcloud\Endpoint\NxWebdavEndpoint: {}
+  Drupal\poc_nextcloud\EntityObserver\NextcloudUsers:
+    arguments:
+      # Explicitly set arguments that cannot be autowired.
+      # See https://www.drupal.org/project/drupal/issues/3336739
+      Psr\Log\LoggerInterface: '@logger.channel.poc_nextcloud'
+  poc_nextcloud.entity_observer:
+    class: Drupal\poc_nextcloud\EntityObserver\Multicast
+    factory: [~, fromServiceIdsWithEmptyFallback]
```

Comment:
There are different types of factories:
- Reusable factories for a range of service instances. Here, having a dedicated factory service allows to separate the parameter signatures: The factory service gets the services that are needed as dependencies, whereas the factory method only gets the parameters that distinguish two instances from each other. This way additional instances can be added without repeating the main dependencies argument list.
- Alternative constructors: These are usually meant for a single instance. Here a static method is ok. But sometimes the original class lives in a different package or is native php code, so the factory has to be outside the class. E.g. SplString for site.path.factory, or guzzle client.
- Factory meant to be called from all over the place: E.g. Database::getConnection(). These factories are not only called once from the container, but from anywhere in the application where dependency injection is not fully available. These factories need to be easy and short, and it cannot ask for all the dependencies as parameters. Instead it will just call container::get()

In core.services.yml a lot of the non-static factories fall into the first category.